Implement font-size: math.

eerii opened the pull request:
#### 9d77743bb040527f2436a48da3c7c6a179e4ef0d
<pre>
Implement font-size: math.
<a href="https://bugs.webkit.org/show_bug.cgi?id=49309">https://bugs.webkit.org/show_bug.cgi?id=49309</a>

Reviewed by Frédéric Wang.

Parse `font-size: math` and add the logic to compute the
scale of the font <a href="https://www.w3.org/TR/css-fonts-4/#valdef-font-size-math.">https://www.w3.org/TR/css-fonts-4/#valdef-font-size-math.</a>
This is gated behind the `CSSMathDepth` preference.
A new user agent spreadsheet is added to keep the legacy behaviour when
the preference is not enabled.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/math-script-level-and-math-style/math-script-level-002.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/math-script-level-and-math-style/math-script-level-004.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/mathml/relations/css-styling/scriptlevel-001-expected.txt: Renamed from LayoutTests/platform/ios/imported/w3c/web-platform-tests/mathml/relations/css-styling/scriptlevel-001-expected.txt.
* LayoutTests/imported/w3c/web-platform-tests/mathml/relations/css-styling/size-containment-001.tentative-expected.txt: Renamed from LayoutTests/platform/glib/imported/w3c/web-platform-tests/mathml/relations/css-styling/size-containment-001.tentative-expected.txt.
* LayoutTests/mathml/presentation/scripts-subsup-expected.html:
* LayoutTests/mathml/presentation/scripts-subsup.html:
* LayoutTests/mathml/presentation/scripts-underover-expected.html:
* LayoutTests/mathml/presentation/scripts-underover.html:
* LayoutTests/platform/glib/TestExpectations:
* LayoutTests/platform/glib/accessibility/math-multiscript-attributes-expected.txt:
* LayoutTests/platform/glib/mathml/opentype/horizontal-expected.txt:
* LayoutTests/platform/glib/mathml/opentype/horizontal-munderover-expected.txt:
* LayoutTests/platform/glib/mathml/opentype/opentype-stretchy-horizontal-expected.txt:
* LayoutTests/platform/glib/mathml/presentation/bug159513-expected.txt:
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/mathml/relations/css-styling/scriptlevel-001-expected.txt:
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/mathml/relations/css-styling/size-containment-001.tentative-expected.txt:
* LayoutTests/platform/ios/mathml/opentype/opentype-stretchy-horizontal-expected.txt:
* LayoutTests/platform/mac/imported/w3c/web-platform-tests/mathml/relations/css-styling/scriptlevel-001-expected.txt: Removed.
* LayoutTests/platform/mac/imported/w3c/web-platform-tests/mathml/relations/css-styling/size-containment-001.tentative-expected.txt:
* LayoutTests/platform/mac/mathml/opentype/opentype-stretchy-horizontal-expected.txt:
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/CSSValueKeywords.in:
* Source/WebCore/css/mathml.css:
(mroot &gt; :not(:first-child)):
(msub &gt; :not(:first-child),):
* Source/WebCore/css/mathmlFontSizeMath.css: Added.
(#if defined(ENABLE_MATHML) &amp;&amp; ENABLE_MATHML):
(math):
* Source/WebCore/css/mathmlLegacyFontSizeMath.css: Added.
(#if defined(ENABLE_MATHML) &amp;&amp; ENABLE_MATHML):
(msub &gt; :not(:first-child),):
* Source/WebCore/css/parser/CSSParserContext.cpp:
(WebCore::applyUASheetBehaviorsToContext):
(WebCore::add):
* Source/WebCore/css/parser/CSSParserContext.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Font.cpp:
(WebCore::CSSPropertyParserHelpers::consumeFontSizeUnresolved):
* Source/WebCore/style/InspectorCSSOMWrappers.cpp:
(WebCore::Style::InspectorCSSOMWrappers::collectDocumentWrappers):
* Source/WebCore/style/StyleBuilderCustom.h:
(WebCore::Style::BuilderCustom::determineMathDepthScale):
(WebCore::Style::BuilderCustom::applyValueFontSize):
* Source/WebCore/style/StyleResolveForFont.cpp:
(WebCore::Style::fontSizeFromUnresolvedFontSize):
* Source/WebCore/style/UserAgentStyle.cpp:
(WebCore::Style::UserAgentStyle::ensureDefaultStyleSheetsForElement):
* Source/WebCore/style/UserAgentStyle.h:

Canonical link: <a href="https://commits.webkit.org/304802@main">https://commits.webkit.org/304802@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/37465a6484f54f01a62e77bc7478a09c78b8f1e8

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/136262 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/8619 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/47542 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/143973 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/89233 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/138133 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/9297 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/8463 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/104187 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/71996 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/139207 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/6760 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/122098 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/85019 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/6416 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/4076 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/4565 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/115705 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/40299 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/146717 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/8301 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/40866 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/112526 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/8318 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/6973 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/112869 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28728 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/6344 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/118403 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/62288 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/8349 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/36460 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/8067 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/71908 "Built successfully") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/8289 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/8141 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

fred-wang commented on Source/WebCore/platform/graphics/opentype/OpenTypeMathData.cpp:
```diff
@@ -303,6 +303,17 @@ float OpenTypeMathData::getMathConstant(const Font& font, MathConstant constant)
+        case ScriptScriptPercentScaleDown:
+            return 0.5041;
+        default:
+            // FIXME: Add fallbacks for other constants
```
ScriptPercentScaleDown/ScriptScriptPercentScaleDown are treated specially when they are null https://w3c.github.io/mathml-core/#layout-constants-mathconstants but I think that's not the case for most constants.

I guess it's fine to handle ScriptPercentScaleDown/ScriptScriptPercentScaleDown here but I wonder if it should just be handled in BuilderCustom::determineMathDepthScale.

eerii replied:
I have moved them to `determineMathDepthScale`

eerii commented on Source/WebCore/css/parser/CSSParserContext.cpp:
```diff
@@ -118,13 +119,14 @@ CSSParserContext::CSSParserContext(const Document& document, const URL& sheetBas
 void add(Hasher& hasher, const CSSParserContext& context)
 {
-    uint32_t bits = context.isHTMLDocument                  << 0
+    uint64_t bits = context.isHTMLDocument                  << 0
```
I'm a bit unsure about this change. We need to add the `cssMathDepthEnabled` flag here to be able to use it in the parser (specifically for `math@settings-flag=...`), but we are already at the size limit of the type. Increasing it like so works, however, does this create issues or could it be a performance hit?

fred-wang replied:
basically I think it should be ok, it can probably change when hash collision and so force unnecessary invalidation, but I'm not sure what we can do here.

It seems WTF::add can take an arbitrary number of arg, so maybe you can also add an extra uint32_t argument instead, but not sure that will make a difference in term of hash collision.

eerii replied:
I went with adding a second `uint32_t` argument since that avoids writing the cast for each `context.*` member. Thanks

eerii commented on Source/WebCore/style/StyleResolveForFont.cpp:
```diff
@@ -241,6 +241,12 @@ static ResolvedFontSize fontSizeFromUnresolvedFontSize(const CSSPropertyParserHe
                     .keyword = CSSValueInvalid
                 };
 
+            case CSSValueMath:
+                return {
+                    .size = 0.0f,
+                    .keyword = CSSValueInvalid
+                };
```
Added this fix for the `2d.text.setFont.mathFont` tests. If I understand correctly this is used for canvas, and that not having it causes a crash. I followed [Chromium's patch](https://chromium-review.googlesource.com/c/chromium/src/+/3194662) to set the size to 0 here.

fred-wang commented on LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/mathml/relations/css-styling/writing-mode/writing-mode-002-expected.txt:
```diff
@@ -0,0 +1,31 @@
+
+PASS Layout of maction (writing-mode: horizontal-tb; direction: rtl;)
+FAIL Layout of menclose (writing-mode: horizontal-tb; direction: rtl;) assert_approx_equals: inline position (child 0) expected 7.96875 +/- 1 but got 3.1875
```
This test seems to assume menclose is treated as an mrow, maybe it should force MathMLCore to be enabled, so we just pass it.

eerii replied:
Even if we force MathMLCore, WebKit still renders the menclose like a menclose and not a mrow. We could change that making it an unknown element, but based on [this discussion](https://github.com/w3c/mathml-core/issues/245) there seems to be an intention of adding it to Core eventually?

Fwiw this test case was [already failing](https://wpt.fyi/results/mathml/relations/css-styling/writing-mode/writing-mode-002.html) for both Gecko and WebKit (since they render the menclose) so I think any change can be made outside the scope of this patch.

fred-wang replied:
Oh right. I forgot we still render menclose.

fred-wang commented on LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/mathml/relations/html5-tree/dynamic-childlist-002-expected.txt:
```diff
@@ -0,0 +1,28 @@
+PASS Appending and removing children to mrow
+PASS Appending and removing children to ms
+PASS Appending and removing children to mspace
+FAIL Appending and removing children to msqrt assert_approx_equals: inline size expected 13 +/- 1 but got 16
```
Why do we fail that one now?

eerii replied:
It's not that we fail that (it was already failing), it's that WK1 started passing the first and last ones like glib. Not really sure why it is passing those and WK2 is not though. However, it may be different after this test run as I think your comment on the stylesheets was accurate.

EDIT: After fixing the stylesheet now the other tests also pass in WK2.

fred-wang commented on Source/WebCore/css/mathml.css:
```diff
@@ -194,7 +194,7 @@ mfrac > :nth-child(2) {
 /* Other rules for scriptlevel, displaystyle and math-shift */
 mroot > :not(:first-child) {
-    font-size: 0.5041em; /* This 0.71^2 since the scriptlevel is incremented by 2 in the index. */
+    font-size: 0.5041em; /* Overwritten by mathMLFontSizeMath.css when CSSMathDepthEnabled is on */
```
The rule in mathMLFontSizeMath.css seems of lower selector specificity, so can you explain why WebKit would override it?

I wonder if we should just move these legacy rules in a separate stylesheet too, so it's clear which rule is taken depending on whether the flag is on or off...

fred-wang replied:
Also see https://webkit.org/code-style-guidelines/#comments-sentences about the comment format that was previously used.

eerii replied:
I don't think they were being overwritten, you are right, sorry. I separated the legacy rules into a different stylesheet so now it should be ok.

It would be very nice to have something similar to gecko that allows gating preferences inside the same CSS stylesheet.

fred-wang commented on Source/WebCore/css/mathml.css:
```diff
@@ -210,7 +210,7 @@ mmultiscripts > :not(:first-child),
 mover > :not(:first-child),
 munderover > :not(:first-child) {
-    font-size: 0.71em;
+    font-size: 0.71em; /* Overwritten by mathMLFontSizeMath.css when CSSMathDepthEnabled is on */
```
Ditto.

fred-wang commented on Source/WebCore/css/mathmlFontSizeMath.css:
```diff
@@ -0,0 +1,36 @@
+ */
+
+/* https://w3c.github.io/mathml-core/#user-agent-stylesheet */
+/* This file contains legacy styles to be applied when CSSMathDepthEnabled is on */
```
nit missing period, see https://webkit.org/code-style-guidelines/#comments-sentences

fred-wang reviewed (approved):
I thought I had left comments the other day, but not sure it worked.

anttijk commented on Source/WebCore/css/mathmlFontSizeMath.css:
```diff
@@ -0,0 +1,36 @@
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/* https://w3c.github.io/mathml-core/#user-agent-stylesheet */
+/* This file contains the styles that should be gated by the preference CSSMathDepthEnabled. */
+
+/* Universal rules */
+* {
+  font-size: math;
+}
```
Since there is no namespace this is saying that every element in the document has `font-size:math`?

fred-wang replied:
The MathML Core stylesheet has `@namespace "http://www.w3.org/1998/Math/MathML";` it should only apply to MathML elements. @eerii can you check if that's the case with the way you split the stylesheets? If not, I would expect many existing tests to fail...

eerii replied:
That was an oversight in creating the conditional stylesheets, thank you for noticing. It is also an issue in the other stylesheets, I'll file a separate PR for those. This explains some of the test failures I was seeing.

anttijk commented on LayoutTests/inspector/css/shadow-scoped-style.html:
```diff
@@ -1,4 +1,4 @@
-<!DOCTYPE html>
+<!DOCTYPE html><!-- webkit-test-runner [ CSSMathDepthEnabled=false ] -->
```
why is there a bunch of tests where this feature is turned off?

eerii replied:
These are not needed anymore, the tests seem to be working ok now with the correct namespace (at least locally on GTK, let's wait for the EWS results).

fred-wang commented on Source/WebCore/css/mathmlLegacyFontSizeMath.css:
```diff
@@ -0,0 +1,41 @@
+/* This file contains the legacy styles that should be applied if CSSMathDepthEnabled if disabled. */
+
+/* Other rules for scriptlevel, displaystyle and math-shift */
+mroot > :not(:first-child) {
```
the same comment about namespace applies.

fred-wang commented on LayoutTests/imported/w3c/web-platform-tests/mathml/relations/css-styling/writing-mode/writing-mode-002.html:
```diff
@@ -1,4 +1,4 @@
-<!DOCTYPE html>
+<!DOCTYPE html><!-- webkit-test-runner [ CoreMathMLEnabled=true ] -->
```
why is it needed? i thought you mentioned this does not change the behavior for  menclose anyway. and the test does not involve font-size: math?

eerii replied:
Hmm, I think that might have been leftover from a previous iteration of this patch. I tried removing it locally and it passes.

fred-wang commented on LayoutTests/platform/mac-wk1/TestExpectations:
```diff
@@ -3263,6 +3263,8 @@ imported/w3c/web-platform-tests/css/css-images/gradient/conic-gradient-001.html
 
 webkit.org/b/295119  imported/w3c/web-platform-tests/css/css-fonts/font-size-adjust-text-orientation.html [ Pass ImageOnlyFailure ]
 
+imported/w3c/web-platform-tests/css/css-fonts/math-script-level-and-math-style/font-size-math-002.tentative.html [ ImageOnlyFailure ]
```
do we know why it's failing? can you please open a follow-up bug with a a webkit.org/b/XXXXX annotation here?

eerii replied:
That was a tricky one to figure out. It was very strange because it was only affecting WK1, and the size was way smaller than it needed to be, but this property shouldn't have any different between WK1/2. I have removed it to test with EWS again, let's see what the issue is now that other problems are resolved such as the namespace. I'll open up a bug if it persists.

fred-wang commented on LayoutTests/mathml/presentation/scripts-subsup.html:
```diff
@@ -1,28 +1,26 @@
 <!doctype html>
-<meta name="fuzzy" content="maxDifference=187-255; totalPixels=213-318" />
+<meta name="fuzzy" content="maxDifference=0-255; totalPixels=0-520" />
```
Which such big fuzzy tolerance, I wonder how relevant the test is now... 

Ideally, if we can make these tests more robust (some suggestions below) we could probably export them to WPT. Or maybe we have equivalent/better tests in WPT already and we should just remove these.

fred-wang commented on LayoutTests/mathml/presentation/scripts-subsup.html:
```diff
@@ -1,28 +1,26 @@
+      <!-- This verifies that the scripts are drawn smaller. -->
       <math>
         <msub>
           <mi>A</mi>
```
Could you instead set `font-family: Ahem` on the `<math>` tag and use `<mn>` elements instead, so that we obtain black squares that are likely less subject to fuzziness? Or alternatively `<mspace>` elements whose sizes are set in `em` (so font-size dependent).

fred-wang commented on LayoutTests/mathml/presentation/scripts-subsup.html:
```diff
@@ -1,28 +1,26 @@
         <msub>
           <mi>A</mi>
-          <mi style="font-size: 0.75em !important">B</mi>
+          <mi style="font-size: 0.71em !important">B</mi>
```
nit: This is already in the current test, but I believe `"font-size: 0.71em !important"` should really be set on the reference file (`-expected.html`) not this one.

fred-wang commented on LayoutTests/mathml/presentation/scripts-underover-expected.html:
```diff
@@ -17,6 +17,9 @@
       margin: 0;
       }
+      math {
+      font-family: Ahem;
```
nit: indentation

fred-wang commented on LayoutTests/mathml/presentation/scripts-underover-expected.html:
```diff
@@ -112,19 +115,20 @@
     </table>
 
     <div style="position: absolute; top: 250px; left: 10px;">
+      <!-- This verifies that the scripts are drawn smaller. -->
```
nit: maybe you can explain a bit more that this is a 0.71 scale factor because that's the default value suggested by the spec when the current font does not have a MATH table.

(If we upstream that, it's likely this won't work well in Firefox, because it will try to pick a math font)

eerii replied:
I have tested this on Firefox and while it does work for me, you are right that it may not always work depending on the font configuration. We should also probably split this test if we upstream it because the two parts (the table and the small scripts) are not that related and have different fuzzing ranges.

fred-wang commented on LayoutTests/mathml/presentation/scripts-underover.html:
```diff
@@ -18,6 +18,9 @@
       margin: 0;
       }
+      math {
+      font-family: Ahem;
```
nit: indent

eerii committed:
Implement font-size: math.
https://bugs.webkit.org/show_bug.cgi?id=49309

Reviewed by Frédéric Wang.

Parse `font-size: math` and add the logic to compute the
scale of the font https://www.w3.org/TR/css-fonts-4/#valdef-font-size-math.
This is gated behind the `CSSMathDepth` preference.
A new user agent spreadsheet is added to keep the legacy behaviour when
the preference is not enabled.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/math-script-level-and-math-style/math-script-level-002.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/math-script-level-and-math-style/math-script-level-004.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/mathml/relations/css-styling/scriptlevel-001-expected.txt: Renamed from LayoutTests/platform/ios/imported/w3c/web-platform-tests/mathml/relations/css-styling/scriptlevel-001-expected.txt.
* LayoutTests/imported/w3c/web-platform-tests/mathml/relations/css-styling/size-containment-001.tentative-expected.txt: Renamed from LayoutTests/platform/glib/imported/w3c/web-platform-tests/mathml/relations/css-styling/size-containment-001.tentative-expected.txt.
* LayoutTests/mathml/presentation/scripts-subsup-expected.html:
* LayoutTests/mathml/presentation/scripts-subsup.html:
* LayoutTests/mathml/presentation/scripts-underover-expected.html:
* LayoutTests/mathml/presentation/scripts-underover.html:
* LayoutTests/platform/glib/TestExpectations:
* LayoutTests/platform/glib/accessibility/math-multiscript-attributes-expected.txt:
* LayoutTests/platform/glib/mathml/opentype/horizontal-expected.txt:
* LayoutTests/platform/glib/mathml/opentype/horizontal-munderover-expected.txt:
* LayoutTests/platform/glib/mathml/opentype/opentype-stretchy-horizontal-expected.txt:
* LayoutTests/platform/glib/mathml/presentation/bug159513-expected.txt:
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/mathml/relations/css-styling/scriptlevel-001-expected.txt:
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/mathml/relations/css-styling/size-containment-001.tentative-expected.txt:
* LayoutTests/platform/ios/mathml/opentype/opentype-stretchy-horizontal-expected.txt:
* LayoutTests/platform/mac/imported/w3c/web-platform-tests/mathml/relations/css-styling/scriptlevel-001-expected.txt: Removed.
* LayoutTests/platform/mac/imported/w3c/web-platform-tests/mathml/relations/css-styling/size-containment-001.tentative-expected.txt:
* LayoutTests/platform/mac/mathml/opentype/opentype-stretchy-horizontal-expected.txt:
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/CSSValueKeywords.in:
* Source/WebCore/css/mathml.css:
(mroot > :not(:first-child)):
(msub > :not(:first-child),):
* Source/WebCore/css/mathmlFontSizeMath.css: Added.
(#if defined(ENABLE_MATHML) && ENABLE_MATHML):
(math):
* Source/WebCore/css/mathmlLegacyFontSizeMath.css: Added.
(#if defined(ENABLE_MATHML) && ENABLE_MATHML):
(msub > :not(:first-child),):
* Source/WebCore/css/parser/CSSParserContext.cpp:
(WebCore::applyUASheetBehaviorsToContext):
(WebCore::add):
* Source/WebCore/css/parser/CSSParserContext.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Font.cpp:
(WebCore::CSSPropertyParserHelpers::consumeFontSizeUnresolved):
* Source/WebCore/style/InspectorCSSOMWrappers.cpp:
(WebCore::Style::InspectorCSSOMWrappers::collectDocumentWrappers):
* Source/WebCore/style/StyleBuilderCustom.h:
(WebCore::Style::BuilderCustom::determineMathDepthScale):
(WebCore::Style::BuilderCustom::applyValueFontSize):
* Source/WebCore/style/StyleResolveForFont.cpp:
(WebCore::Style::fontSizeFromUnresolvedFontSize):
* Source/WebCore/style/UserAgentStyle.cpp:
(WebCore::Style::UserAgentStyle::ensureDefaultStyleSheetsForElement):
* Source/WebCore/style/UserAgentStyle.h:

Canonical link: https://commits.webkit.org/304802@main
